### PR TITLE
apiserver aggregator upgrade unit test

### PR DESCRIPTION
**Pull request by aojea**
/kind failing-test

Add unit tests to cover regressions against #104985

In order to be able to mock the egress-selector add a new constructor with the following signature

```
// NewEgressSelectorWithMap returns a EgressSelector with the supplied EgressType to DialFunc map.
func NewEgressSelectorWithMap(m map[EgressType]utilnet.DialFunc) (*EgressSelector, error) {
```

The tests verifies that a good EgressSelector doesn't break upgrade requests, and that the EgressSelector is never bypassed by the upgrade logic.

Before 224bcf63f252658fffc418a1d90530c6867f8b3c the apiserver-aggregator bypassed the EgressSelector and connected directly to the endpoint, explained in https://github.com/kubernetes/kubernetes/pull/104985

```
=== RUN   TestProxyUpgrade
=== RUN   TestProxyUpgrade/valid_hostname_+_CABundle
=== RUN   TestProxyUpgrade/invalid_hostname_+_insecure
=== RUN   TestProxyUpgrade/invalid_hostname_+_CABundle
2021/10/05 00:25:58 http: TLS handshake error from 127.0.0.1:54986: remote error: tls: bad certificate
=== RUN   TestProxyUpgrade/valid_hostname_+_CABundle_+_egress_selector
    /home/aojea/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:588: valid hostname + CABundle + egress selector: expected websocket error, got none
    /home/aojea/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:564: valid hostname + CABundle + egress selector: expected called=false, got true
=== RUN   TestProxyUpgrade/invalid_hostname_+_insecure_+_Egress_Selector
    /home/aojea/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:588: invalid hostname + insecure + Egress Selector: expected websocket error, got none
    /home/aojea/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:564: invalid hostname + insecure + Egress Selector: expected called=false, got true
=== RUN   TestProxyUpgrade/invalid_hostname_+_CABundle_+_EgressSelector
2021/10/05 00:25:58 http: TLS handshake error from 127.0.0.1:51748: remote error: tls: bad certificate
--- FAIL: TestProxyUpgrade (0.02s)
    --- PASS: TestProxyUpgrade/valid_hostname_+_CABundle (0.00s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_insecure (0.00s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_CABundle (0.00s)
    --- FAIL: TestProxyUpgrade/valid_hostname_+_CABundle_+_egress_selector (0.00s)
    --- FAIL: TestProxyUpgrade/invalid_hostname_+_insecure_+_Egress_Selector (0.00s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_CABundle_+_EgressSelector (0.00s)
    
```

After 224bcf63f252658fffc418a1d90530c6867f8b3c , the egress selector is not bypassed and fails

```
=== RUN   TestProxyUpgrade
=== RUN   TestProxyUpgrade/invalid_hostname_+_insecure
=== RUN   TestProxyUpgrade/invalid_hostname_+_CABundle
2021/10/05 00:23:30 http: TLS handshake error from 127.0.0.1:47942: remote error: tls: bad certificate
=== RUN   TestProxyUpgrade/valid_hostname_+_CABundle_+_egress_selector
=== RUN   TestProxyUpgrade/invalid_hostname_+_insecure_+_Egress_Selector
=== RUN   TestProxyUpgrade/invalid_hostname_+_CABundle_+_EgressSelector
=== RUN   TestProxyUpgrade/valid_hostname_+_CABundle
--- PASS: TestProxyUpgrade (0.01s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_insecure (0.00s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_CABundle (0.00s)
    --- PASS: TestProxyUpgrade/valid_hostname_+_CABundle_+_egress_selector (0.00s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_insecure_+_Egress_Selector (0.00s)
    --- PASS: TestProxyUpgrade/invalid_hostname_+_CABundle_+_EgressSelector (0.00s)
    --- PASS: TestProxyUpgrade/valid_hostname_+_CABundle (0.00s)
PASS
```

```release-note
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
/sig api-machinery
/assign @liggitt @caesarxuchao 
I was trying to find another way of testing it when I found that previous upgrade bypassed completely the egress selector.

**Comment by caesarxuchao** [comment]
Thanks, @aojea!

The added tests verified that a bad egressSelector can break upgrade requests (note that all the `ExpectError` in the new test cases are set to `true`).

In addition, we need to verify that good egressSelector does not break upgrade requests. Can you take a look at the mock egressSelector I added in this [commit](https://github.com/caesarxuchao/kubernetes/commit/362d2f9369ee1087fad02f8d2274f2d61a56a339) and add that to your test cases?

**Comment by aojea** [comment]
> Thanks, @aojea!
> 
> The added tests verified that a bad egressSelector can break upgrade requests (note that all the `ExpectError` in the new test cases are set to `true`).
> 
> In addition, we need to verify that good egressSelector does not break upgrade requests. Can you take a look at the mock egressSelector I added in this [commit](https://github.com/caesarxuchao/kubernetes/commit/362d2f9369ee1087fad02f8d2274f2d61a56a339) and add that to your test cases?

done, PTAL

**Comment by aojea** [comment]
> Kubernetes e2e suite: [sig-node] PreStop graceful pod terminated should wait until preStop hook completes the process expand_more

/test pull-kubernetes-e2e-gce-ubuntu-containerd

File: staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go [hunk 1]
```diff
@@ -358,6 +358,16 @@ func NewEgressSelector(config *apiserver.EgressSelectorConfiguration) (*EgressSe
...
 }
 
+// NewEgressSelectorWithMap returns a EgressSelector with the supplied EgressType to DialFunc map.
+func NewEgressSelectorWithMap(m map[EgressType]utilnet.DialFunc) (*EgressSelector, error) {
+	if m == nil {
+		m = make(map[EgressType]utilnet.DialFunc)
+	}
+	return &EgressSelector{
+		egressToDialer: m,
+	}, nil
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go:368]: nit: We can remove the `error` from the signature.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go [hunk 2]
```diff
@@ -379,11 +383,42 @@ func TestProxyHandler(t *testing.T) {
...
+
+func newDialerAndSelector() (*mockEgressDialer, *egressselector.EgressSelector, error) {
+	dialer := &mockEgressDialer{}
+	m := make(map[egressselector.EgressType]utilnet.DialFunc)
+	m[egressselector.Cluster] = dialer.dial
+	es, err := egressselector.NewEgressSelectorWithMap(m)
+	return dialer, es, err
+}
+
+func newDialerBrokenAndSelector() (*mockEgressDialer, *egressselector.EgressSelector, error) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:408]: nit: newBrokenDialerAndSelector sounds more natural.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go [hunk 3]
```diff
@@ -436,14 +471,121 @@ func TestProxyUpgrade(t *testing.T) {
 			ExpectError:  true,
 			ExpectCalled: false,
 		},
+		"valid hostname + CABundle + egress selector": {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:474]: I think we only need the "valid hostname + CABundle + egress selector" and the "valid hostname + CABundle + egress selector non working" cases. The cases with invalid hostname or invalid CA do not tell us if the egress selector code works as expected.

Comment (aojea) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:474]: done

File: staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go (hunk 1 above)
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go]: done

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go (hunk 2 above)
Comment (aojea) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go]: done

**Comment by caesarxuchao** [comment]
/lgtm

@liggitt can you also take a look? Thanks.

**Pull request merged**